lanternmq/rabbitmq/messagequeue.go

```diff
@@ -0,0 +1,346 @@
+// kind: "topic"
+// durable: true
+// autoDelete: false
+// intenral: false
+// noWait: false
+// args: nil
+func (mq *MessageQueue) DeclareTopic(chID lanternmq.ChannelID, name string) error {
+	ch, err := mq.getChannel(chID)
+	if err != nil {
+		return err
+	}
+
+	err = ch.ExchangeDeclare(
+		name,    // name
+		"topic", // type
+		true,    // durable
+		false,   // auto-deleted
+		false,   // internal
+		false,   // no-wait
+		nil,     // arguments
+	)
+	if err != nil {
+		err = errors.New("unable to declare target")
+	}
+
+	return err
+}
+
+// PublishToTopic sends 'message' to the topic 'name' over the channel with ID 'chID' with routing key 'routingKey'. It
+// uses RabbitMQ's Publish method with the following arguments:
+// exchange: name
+// key: routingKey
+// mandatory: false
+// immediate: false
+// publishing:
+//   ContentType: "text/plain"
+//   Body: []byte(message)
+func (mq *MessageQueue) PublishToTopic(chID lanternmq.ChannelID, name string, routingKey string, message string) error {
+	ch, err := mq.getChannel(chID)
+	if err != nil {
+		return err
+	}
+
+	err = ch.Publish(
+		name,       // exchange
+		routingKey, // routing key
+		false,      // mandatory
+		false,      // immediate
+		amqp.Publishing{
+			ContentType: "text/plain",
+			Body:        []byte(message),
+		})
+	if err != nil {
+		err = fmt.Errorf("unable to publish to target %s with routing key %s", name, routingKey)
+	}
+
+	return err
+}
+
+// DeclareTopicReceiveQueue creates a queue named 'qName' to receive messages from the topic named 'topicName'
+// with routing key 'routingKey' over the channel with ID 'chID'. It uses the RabbitMQ method QueueDeclare with
+// the following arguments:
+// name: qName
+// durable: false
+// autoDelete: false
+// exclusive: true
+// noWait: false
+// args: nil
+//
+// It then calls the RabbitMQ method QueueBind with the following arguments:
+// name: qName
+// key: routingKey
+// exchange: topicName
+// noWait: false
+// args: nil
+func (mq *MessageQueue) DeclareTopicReceiveQueue(chID lanternmq.ChannelID, topicName string, qName string, routingKey string) error {
+	ch, err := mq.getChannel(chID)
+	if err != nil {
+		return err
+	}
+
+	_, err = ch.QueueDeclare(
+		qName, // name
+		false, // durable
+		false, // delete when usused
+		true,  // exclusive
+		false, // no-wait
+		nil,   // arguments
+	)
+	if err != nil {
+		err = fmt.Errorf("unable to create queue: %s", err.Error())
+		return err
+	}
+
+	err = ch.QueueBind(
+		qName,      // queue name
+		routingKey, // routing key
+		topicName,  // exchange
+		false,
+		nil)
+	if err != nil {
+		err = fmt.Errorf("unable to bind queue %s to target %s with routing key %s", qName, topicName, routingKey)
+		return err
+	}
+
+	return err
+}
+
+// Close closes each channel that's been created, and then closes the connection to the underlying RabbitMQ
+// message service.
+func (mq *MessageQueue) Close() {
+	if mq.channels != nil {
+		for _, ch := range mq.channels {
+			ch.Close()
+		}
```

Comment:
At this point, the channel IDs in the mq.channels slice are invalid.  May want to consider removing the elements from the slice after closing each channel.